kubernetes-informer/src/main/java/io/micronaut/kubernetes/client/informer/ResourceEventHandlerConstructorInterceptor.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.kubernetes.client.informer;
+
+import io.kubernetes.client.Discovery;
+import io.kubernetes.client.common.KubernetesObject;
+import io.kubernetes.client.informer.ResourceEventHandler;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.informer.SharedInformerFactory;
+import io.kubernetes.client.util.Namespaces;
+import io.micronaut.aop.ConstructorInterceptor;
+import io.micronaut.aop.ConstructorInvocationContext;
+import io.micronaut.aop.InterceptorBean;
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.kubernetes.client.DiscoveryCache;
+import io.micronaut.kubernetes.client.NamespaceResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Constructor interceptor for the {@link ResourceEventHandler} beans annotated by {@link Informer} annotation that
+ * based on provided parameters in the {@link Informer} annotation created the {@link SharedIndexInformer} and registers
+ * {@link ResourceEventHandler} to the informer.
+ *
+ * @param <ApiType> type of Kubernetes Object
+ * @author Pavol Gressa
+ * @since 3.1
+ */
+@Requires(beans = SharedInformerFactory.class)
+@InterceptorBean(Informer.class)
+@Internal
+public class ResourceEventHandlerConstructorInterceptor<ApiType extends KubernetesObject> implements ConstructorInterceptor<ResourceEventHandler<ApiType>> {
```

Comment:
Note: Constructor interceptor that for every @Informer annotated ResourceEventHandler class will create SharedInformer.